File: web/src/__tests__/editor/MoveLineLogic.test.ts
import { describe, it, expect } from 'vitest'
import { LineState } from '../../editor/LineState'
import { getIndentLevel, getLogicalBlock } from '../../editor/IndentationUtils'
import { findMoveUpTarget, findMoveDownTarget, getMoveTarget, wouldOrphanChildren } from '../../editor/MoveTargetFinder'
import { calculateMove } from '../../editor/MoveExecutor'
import { getSelectedLineRange } from '../../editor/SelectionCoordinates'
import { SELECTION_NONE } from '../../editor/EditorSelection'

function makeLines(...texts: string[]): LineState[] {
  return texts.map((t) => new LineState(t))
}

describe('getIndentLevel', () => {
  it('plain text has indent 0', () => {
    const lines = makeLines('hello')
    expect(getIndentLevel(lines, 0)).toBe(0)
  })

  it('empty text has indent 0', () => {
    const lines = makeLines('')
    expect(getIndentLevel(lines, 0)).toBe(0)
  })

  it('tabs give indent level', () => {
    const lines = makeLines('\t\thello')
    expect(getIndentLevel(lines, 0)).toBe(2)
  })

  it('tab + bullet', () => {
    const lines = makeLines('\t\u2022 hello')
    expect(getIndentLevel(lines, 0)).toBe(1)
  })

  it('invalid index returns 0', () => {
    const lines = makeLines('hello')
    expect(getIndentLevel(lines, 5)).toBe(0)
  })
})

describe('getLogicalBlock', () => {
  it('single line block', () => {
    const lines = makeLines('a', 'b', 'c')
    expect(getLogicalBlock(lines, 1)).toEqual([1, 1])
  })

  it('includes deeper children', () => {
    const lines = makeLines('parent', '\tchild1', '\tchild2', 'sibling')
    expect(getLogicalBlock(lines, 0)).toEqual([0, 2])
  })

  it('nested blocks', () => {
    const lines = makeLines('a', '\tb', '\t\tc', '\td', 'e')
    expect(getLogicalBlock(lines, 0)).toEqual([0, 3])
  })

  it('stops at same indent level', () => {
    const lines = makeLines('a', '\tb', 'c', '\td')
    expect(getLogicalBlock(lines, 0)).toEqual([0, 1])
  })

  it('handles empty lines in block', () => {
    // Empty lines have indent 0, so they break the block
    const lines = makeLines('a', '\tb', '', '\td')
    expect(getLogicalBlock(lines, 0)).toEqual([0, 1])
  })
})

describe('getSelectedLineRange', () => {
  it('no selection returns focused line', () => {
    const lines = makeLines('a', 'b', 'c')
    expect(getSelectedLineRange(lines, SELECTION_NONE, 1)).toEqual([1, 1])
  })

  it('returns range of selected lines', () => {
    const lines = makeLines('abc', 'def', 'ghi')
    // Select from middle of line 0 to middle of line 2: "bc\ndef\ng"
    expect(getSelectedLineRange(lines, { start: 1, end: 9 }, 0)).toEqual([0, 2])
  })

  it('adjusts end line when selection ends at line boundary', () => {
    const lines = makeLines('abc', 'def', 'ghi')
    // Select "abc\n" = 0 to 4, end is at start of line 1
    expect(getSelectedLineRange(lines, { start: 0, end: 4 }, 0)).toEqual([0, 0])
  })
})

describe('findMoveUpTarget', () => {
  it('returns null when at first line', () => {
    const lines = makeLines('a', 'b', 'c')
    expect(findMoveUpTarget(lines, 0, 0)).toBeNull()
  })

  it('simple move up', () => {
    const lines = makeLines('a', 'b', 'c')
    expect(findMoveUpTarget(lines, 1, 1)).toBe(0)
  })

  it('hops over block', () => {
    const lines = makeLines('a', '\tb', '\tc', 'd')
    // Moving 'd' (index 3) up should hop over block [0,2]
    expect(findMoveUpTarget(lines, 3, 3)).toBe(0)
  })
})

describe('findMoveDownTarget', () => {
  it('returns null when at last line', () => {
    const lines = makeLines('a', 'b', 'c')
    expect(findMoveDownTarget(lines, 2, 2)).toBeNull()
  })

  it('simple move down', () => {
    const lines = makeLines('a', 'b', 'c')
    expect(findMoveDownTarget(lines, 0, 0)).toBe(2) // insert after line 1
  })

  it('hops over block', () => {
    const lines = makeLines('a', 'b', '\tc', '\td')
    // Moving 'a' (index 0) down should hop over block at line 1 [1,3]
    expect(findMoveDownTarget(lines, 0, 0)).toBe(4)
  })
})

describe('getMoveTarget', () => {
  it('with selection uses selected range', () => {
    const lines = makeLines('a', 'b', 'c')
    const sel = { start: 0, end: 1 } // selects line 0
    const target = getMoveTarget(lines, true, sel, 0, false)
    expect(target).not.toBeNull()
  })

  it('without selection uses logical block', () => {
    const lines = makeLines('a', '\tb', 'c')
    const target = getMoveTarget(lines, false, SELECTION_NONE, 0, false)
    // Block is [0,1], move down target should be after 'c'
    expect(target).toBe(3)
  })
})

describe('wouldOrphanChildren', () => {
  it('returns false without selection', () => {
    const lines = makeLines('a', '\tb', 'c')
    expect(wouldOrphanChildren(lines, false, SELECTION_NONE, 0)).toBe(false)
  })

  it('returns true when selection excludes children', () => {
    const lines = makeLines('parent', '\tchild1', '\tchild2')
    // Select only parent (line 0), children would be orphaned
    const sel = { start: 0, end: 6 } // "parent"
    expect(wouldOrphanChildren(lines, true, sel, 0)).toBe(true)
  })

  it('returns false when selection includes all children', () => {
    const lines = makeLines('parent', '\tchild1', '\tchild2')
    // Select all three lines
    const sel = { start: 0, end: 100 }
    expect(wouldOrphanChildren(lines, true, sel, 0)).toBe(false)
  })

  it('returns true when first selected line is not shallowest', () => {
    const lines = makeLines('a', '\tb', 'c')
    // Select only \tb (indent 1), but 'c' (indent 0) is shallowest
    // Actually, selecting just line 1
    const lineStartLine1 = 2 // "a\n" = 2
    const sel = { start: lineStartLine1, end: lineStartLine1 + 2 } // "\tb"
    expect(wouldOrphanChildren(lines, true, sel, 1)).toBe(false) // no children after
  })
})

describe('calculateMove (moveLinesInternal)', () => {
  it('single line move up', () => {
    const lines = makeLines('a', 'b', 'c')
    const result = calculateMove(lines, 1, 1, 0, 1, null)
    expect(result).not.toBeNull()
    expect(result!.newLines).toEqual(['b', 'a', 'c'])
  })

  it('single line move down', () => {
    const lines = makeLines('a', 'b', 'c')
    const result = calculateMove(lines, 0, 0, 2, 0, null)
    expect(result).not.toBeNull()
    expect(result!.newLines).toEqual(['b', 'a', 'c'])
  })

  it('block move up', () => {
    const lines = makeLines('a', 'b', '\tc')
    const result = calculateMove(lines, 1, 2, 0, 1, null)
    expect(result).not.toBeNull()
    expect(result!.newLines).toEqual(['b', '\tc', 'a'])
  })

  it('block move down', () => {
    const lines = makeLines('a', '\tb', 'c')
    const result = calculateMove(lines, 0, 1, 3, 0, null)
    expect(result).not.toBeNull()
    expect(result!.newLines).toEqual(['c', 'a', '\tb'])
  })

  it('invalid range returns null', () => {
    const lines = makeLines('a', 'b')
    expect(calculateMove(lines, -1, 0, 1, 0, null)).toBeNull()
    expect(calculateMove(lines, 0, 5, 1, 0, null)).toBeNull()
  })

  it('target within source returns null', () => {
    const lines = makeLines('a', 'b', 'c')
    expect(calculateMove(lines, 0, 2, 1, 0, null)).toBeNull()
  })

  it('adjusts focused line when within source', () => {
    const lines = makeLines('a', 'b', 'c')
    const result = calculateMove(lines, 1, 1, 0, 1, null)
    expect(result).not.toBeNull()
    expect(result!.newFocusedLineIndex).toBe(0)
  })

  it('adjusts focused line when outside source', () => {
    const lines = makeLines('a', 'b', 'c')
    // Move line 0 to after line 2 (target=3), focused on line 1
    const result = calculateMove(lines, 0, 0, 3, 1, null)
    expect(result).not.toBeNull()
    // Line 1 (b) moves to index 0
    expect(result!.newFocusedLineIndex).toBe(0)
  })

  it('updates selection during move', () => {
    const lines = makeLines('abc', 'def', 'ghi')
    const sel = { start: 0, end: 3 } // selects "abc"
    const result = calculateMove(lines, 0, 0, 2, 0, sel)
    expect(result).not.toBeNull()
    expect(result!.newSelection).not.toBeNull()
  })
})

describe('findMoveUpTarget with hiddenIndices', () => {
  it('skips hidden lines when scanning backward', () => {
    // Lines: a, ☑ done, b  — hide index 1
    const lines = makeLines('a', '☑ done', 'b')
    const hidden = new Set([1])
    expect(findMoveUpTarget(lines, 2, 2, hidden)).toBe(0)
  })

  it('returns null when all lines above are hidden', () => {
    const lines = makeLines('title', '☑ done1', '☑ done2', 'active')
    const hidden = new Set([1, 2])
    // Line 3 trying to move up, lines 1-2 are hidden, line 0 is at same indent (valid)
    expect(findMoveUpTarget(lines, 3, 3, hidden)).toBe(0)
  })

  it('returns null when only line 0 remains and target is hidden', () => {
    const lines = makeLines('☑ done', 'active')
    const hidden = new Set([0])
    expect(findMoveUpTarget(lines, 1, 1, hidden)).toBeNull()
  })

  it('skips hidden block including children', () => {
    // Lines: a, ☑ parent, \tchild, b — hide indices 1,2
    const lines = makeLines('a', '☑ parent', '\tchild', 'b')
    const hidden = new Set([1, 2])
    expect(findMoveUpTarget(lines, 3, 3, hidden)).toBe(0)
  })
})

describe('findMoveDownTarget with hiddenIndices', () => {
  it('skips hidden lines when scanning forward', () => {
    const lines = makeLines('a', '☑ done', 'b')
    const hidden = new Set([1])
    expect(findMoveDownTarget(lines, 0, 0, hidden)).toBe(3)
  })

  it('returns null when all lines below are hidden', () => {
    const lines = makeLines('active', '☑ done1', '☑ done2')
    const hidden = new Set([1, 2])
    expect(findMoveDownTarget(lines, 0, 0, hidden)).toBeNull()
  })

  it('skips hidden block and finds target after it', () => {
    const lines = makeLines('a', '☑ parent', '\tchild', 'b')
    const hidden = new Set([1, 2])
    // Moving 'a' down should skip hidden 1,2 and land after 'b'
    expect(findMoveDownTarget(lines, 0, 0, hidden)).toBe(4)
  })
})

describe('findMoveUpTarget skips hidden block above visible target', () => {
  it('jumps past hidden block between visible lines', () => {
    // Lines: unchecked, ☑ done1, ☑ done2, empty, unchecked2
    // Moving unchecked2 up should land above the hidden block, not between it and the empty line
    const lines = makeLines('unchecked1', '☑ done1', '☑ done2', '', 'unchecked2')
    const hidden = new Set([1, 2])
    expect(findMoveUpTarget(lines, 4, 4, hidden)).toBe(1)
  })

  it('does not skip hidden block at deeper indent', () => {
    // Lines: ☐ parent (indent 1), ☑ child (hidden, indent 1), Root (indent 0), ☐ mover (indent 1)
    // Moving mover up should swap with Root, not jump past the hidden child
    const lines = makeLines('\t☐ parent', '\t☑ child', 'Root', '\t☐ mover')
    const hidden = new Set([1])
    expect(findMoveUpTarget(lines, 3, 3, hidden)).toBe(2)
  })

  it('skips hidden block with children above visible target', () => {
    // Lines: unchecked, ☑ parent, \tchild (both hidden), empty, unchecked2
    const lines = makeLines('unchecked1', '☑ parent', '\tchild', '', 'unchecked2')
    const hidden = new Set([1, 2])
    expect(findMoveUpTarget(lines, 4, 4, hidden)).toBe(1)
  })
})

describe('findMoveDownTarget does not expand block through hidden siblings', () => {
  it('stops block expansion at hidden sibling', () => {
    // Lines: unchecked1, unchecked2, ☑ done1, ☑ done2, empty
    // Moving unchecked1 down should swap with unchecked2, not jump past hidden lines
    const lines = makeLines('unchecked1', 'unchecked2', '☑ done1', '☑ done2', '')
    const hidden = new Set([2, 3])
    expect(findMoveDownTarget(lines, 0, 0, hidden)).toBe(2)
  })

  it('still includes visible children in block expansion', () => {
    // Lines: unchecked1, unchecked2, \tchild, ☑ done (hidden), empty
    // Moving unchecked1 down should swap past unchecked2 + its child, stop at hidden sibling
    const lines = makeLines('unchecked1', 'unchecked2', '\tchild', '☑ done', '')
    const hidden = new Set([3])
    expect(findMoveDownTarget(lines, 0, 0, hidden)).toBe(3)
  })
})

describe('getMoveTarget with hiddenIndices', () => {
  it('uses hiddenIndices for move up', () => {
    const lines = makeLines('a', '☑ done', 'b')
    const hidden = new Set([1])
    const target = getMoveTarget(lines, false, SELECTION_NONE, 2, true, hidden)
    expect(target).toBe(0)
  })

  it('uses hiddenIndices for move down', () => {
    const lines = makeLines('a', '☑ done', 'b')
    const hidden = new Set([1])
    const target = getMoveTarget(lines, false, SELECTION_NONE, 0, false, hidden)
    expect(target).toBe(3)
  })
})

describe('empty line handling in moves', () => {
  it('empty lines have indent 0 and affect block boundaries', () => {
    const lines = makeLines('a', '', 'b')
    // Empty line breaks block
    expect(getLogicalBlock(lines, 0)).toEqual([0, 0])
  })
})

File: app/src/test/java/org/alkaline/taskbrain/ui/currentnote/MoveLineLogicTest.kt
package org.alkaline.taskbrain.ui.currentnote

import org.alkaline.taskbrain.ui.currentnote.move.MoveTargetFinder
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class MoveLineLogicTest {

    // ==================== getIndentLevel ====================

    @Test
    fun `getIndentLevel returns 0 for plain text`() {
        val state = EditorState()
        state.updateFromText("hello")
        assertEquals(0, state.getIndentLevel(0))
    }

    @Test
    fun `getIndentLevel returns 0 for empty line`() {
        val state = EditorState()
        state.updateFromText("line1\n\nline3")
        assertEquals(0, state.getIndentLevel(1))
    }

    @Test
    fun `getIndentLevel returns count of leading tabs`() {
        val state = EditorState()
        state.updateFromText("\t\tindented")
        assertEquals(2, state.getIndentLevel(0))
    }

    @Test
    fun `getIndentLevel returns 1 for single tab`() {
        val state = EditorState()
        state.updateFromText("\titem")
        assertEquals(1, state.getIndentLevel(0))
    }

    @Test
    fun `getIndentLevel handles tabs with bullet`() {
        val state = EditorState()
        state.updateFromText("\t• item")
        assertEquals(1, state.getIndentLevel(0))
    }

    @Test
    fun `getIndentLevel returns 0 for invalid index`() {
        val state = EditorState()
        state.updateFromText("line")
        assertEquals(0, state.getIndentLevel(5))
    }

    // ==================== getLogicalBlock ====================

    @Test
    fun `getLogicalBlock returns single line when no children`() {
        val state = EditorState()
        state.updateFromText("line1\nline2\nline3")
        assertEquals(0..0, state.getLogicalBlock(0))
        assertEquals(1..1, state.getLogicalBlock(1))
        assertEquals(2..2, state.getLogicalBlock(2))
    }

    @Test
    fun `getLogicalBlock includes deeper indented children`() {
        val state = EditorState()
        state.updateFromText("parent\n\tchild1\n\tchild2\nsibling")
        assertEquals(0..2, state.getLogicalBlock(0))
    }

    @Test
    fun `getLogicalBlock includes nested children`() {
        val state = EditorState()
        state.updateFromText("parent\n\tchild\n\t\tgrandchild\nsibling")
        assertEquals(0..2, state.getLogicalBlock(0))
        assertEquals(1..2, state.getLogicalBlock(1))
        assertEquals(2..2, state.getLogicalBlock(2))
    }

    @Test
    fun `getLogicalBlock stops at same indent`() {
        val state = EditorState()
        state.updateFromText("item1\n\tchild\nitem2\n\tchild2")
        assertEquals(0..1, state.getLogicalBlock(0))
        assertEquals(2..3, state.getLogicalBlock(2))
    }

    @Test
    fun `getLogicalBlock handles empty lines as indent 0`() {
        val state = EditorState()
        state.updateFromText("parent\n\tchild\n\nsibling")
        // Empty line (indent 0) stops the block
        assertEquals(0..1, state.getLogicalBlock(0))
    }

    // ==================== getSelectedLineRange ====================

    @Test
    fun `getSelectedLineRange returns focused line when no selection`() {
        val state = EditorState()
        state.updateFromText("line1\nline2\nline3")
        state.focusedLineIndex = 1
        assertEquals(1..1, state.getSelectedLineRange())
    }

    @Test
    fun `getSelectedLineRange returns selected lines`() {
        val state = EditorState()
        state.updateFromText("line1\nline2\nline3\nline4")
        state.setSelection(6, 17) // line2 and line3
        assertEquals(1..2, state.getSelectedLineRange())
    }

    @Test
    fun `getSelectedLineRange handles single line selection`() {
        val state = EditorState()
        state.updateFromText("line1\nline2\nline3")
        state.setSelection(6, 11) // just line2
        assertEquals(1..1, state.getSelectedLineRange())
    }

    // ==================== findMoveUpTarget ====================

    @Test
    fun `findMoveUpTarget returns null at first line`() {
        val state = EditorState()
        state.updateFromText("line1\nline2")
        assertNull(state.findMoveUpTarget(0..0))
    }

    @Test
    fun `findMoveUpTarget returns previous line for simple case`() {
        val state = EditorState()
        state.updateFromText("line1\nline2\nline3")
        assertEquals(1, state.findMoveUpTarget(2..2))
    }

    @Test
    fun `findMoveUpTarget hops over sibling block`() {
        val state = EditorState()
        state.updateFromText("sibling1\n\tchild1\nitem\n\tchild")
        // Moving "item" + child (lines 2-3) up should hop to before "sibling1"
        assertEquals(0, state.findMoveUpTarget(2..3))
    }

    @Test
    fun `findMoveUpTarget hops over single deeper line`() {
        val state = EditorState()
        state.updateFromText("parent\n\tchild\nitem")
        // Moving "item" (line 2, indent 0) up - should hop over "parent" block (lines 0-1)
        // because "parent" is at same indent, and "\tchild" is its child
        assertEquals(0, state.findMoveUpTarget(2..2))
    }

    @Test
    fun `findMoveUpTarget hops over entire previous group with children`() {
        val state = EditorState()
        // Three groups, each with a parent and three children
        state.updateFromText("group1\n\tchild1\n\tchild2\n\tchild3\ngroup2\n\tchild4\n\tchild5\n\tchild6\ngroup3\n\tchild7\n\tchild8\n\tchild9")
        // Moving group3 (lines 8-11) up should land at position 4 (before group2)
        assertEquals(4, state.findMoveUpTarget(8..11))
    }

    // ==================== findMoveDownTarget ====================

    @Test
    fun `findMoveDownTarget returns null at last line`() {
        val state = EditorState()
        state.updateFromText("line1\nline2")
        assertNull(state.findMoveDownTarget(1..1))
    }

    @Test
    fun `findMoveDownTarget returns position after next line for simple case`() {
        val state = EditorState()
        state.updateFromText("line1\nline2\nline3")
        // Moving line1 down should go to position 2 (after line2)
        assertEquals(2, state.findMoveDownTarget(0..0))
    }

    @Test
    fun `findMoveDownTarget hops over sibling block`() {
        val state = EditorState()
        state.updateFromText("item\nsibling\n\tchild\nafter")
        // Moving "item" (line 0) down should hop over "sibling" + "child" (lines 1-2)
        assertEquals(3, state.findMoveDownTarget(0..0))
    }

    @Test
    fun `findMoveDownTarget hops over single deeper line`() {
        val state = EditorState()
        state.updateFromText("item\n\tchild\nsibling")
        // Moving "item" (line 0) down - "\tchild" is deeper, just hop over it
        assertEquals(2, state.findMoveDownTarget(0..0))
    }

    // ==================== getMoveTarget ====================

    @Test
    fun `getMoveTarget with selection uses selection range`() {
        val state = EditorState()
        state.updateFromText("line1\nline2\nline3\nline4")
        state.setSelection(6, 17) // lines 1-2
        // Move up should go to position 0
        assertEquals(0, state.getMoveTarget(moveUp = true))
    }

    @Test
    fun `getMoveTarget without selection uses logical block`() {
        val state = EditorState()
        state.updateFromText("item\n\tchild\nsibling")
        state.focusedLineIndex = 0
        // Should get logical block 0..1 and find target
        assertEquals(3, state.getMoveTarget(moveUp = false))
    }

    @Test
    fun `getMoveTarget moves one line when first selected is not shallowest`() {
        val state = EditorState()
        state.updateFromText("parent\n\tchild1\n\tchild2\nsibling")
        // Select "\tchild1\n\tchild2" (lines 1-2) - both at same indent
        state.setSelection(7, 22) // approximate selection
        // First line IS the shallowest (both at indent 1), so normal hopping
        val target = state.getMoveTarget(moveUp = false)
        assertEquals(4, target) // After sibling
    }

    @Test
    fun `getMoveTarget with different indent selection moves one line at a time`() {
        val state = EditorState()
        state.updateFromText("line1\n\tchild\nline2\nline3")
        // Select child + line2 (lines 1-2 with indent 1 and 0)
        // Line 1 has indent 1, line 2 has indent 0 - first is deeper than shallowest
        state.setSelection(6, 18)
        // First line (indent 1) is NOT the shallowest (0), so move one at a time
        assertEquals(0, state.getMoveTarget(moveUp = true)) // Just one line up
    }

    // ==================== wouldOrphanChildren ====================

    @Test
    fun `wouldOrphanChildren returns false without selection`() {
        val state = EditorState()
        state.updateFromText("parent\n\tchild")
        state.focusedLineIndex = 0
        assertFalse(state.wouldOrphanChildren())
    }

    @Test
    fun `wouldOrphanChildren returns false when selection includes all children`() {
        val state = EditorState()
        state.updateFromText("parent\n\tchild1\n\tchild2\nsibling")
        // Select entire logical block
        state.setSelection(0, 22) // parent + children
        assertFalse(state.wouldOrphanChildren())
    }

    @Test
    fun `wouldOrphanChildren returns true when selection excludes children`() {
        val state = EditorState()
        state.updateFromText("parent\n\tchild1\n\tchild2\nsibling")
        // Select just parent (line 0)
        state.setSelection(0, 6)
        assertTrue(state.wouldOrphanChildren())
    }

    @Test
    fun `wouldOrphanChildren returns true when first line is not shallowest`() {
        val state = EditorState()
        state.updateFromText("parent\n\tchild1\n\tchild2\nsibling")
        // Select child1 and child2 (lines 1-2) - first line (indent 1) is not shallowest
        state.setSelection(7, 22)
        // Both at indent 1, so first IS shallowest within selection - should be false
        assertFalse(state.wouldOrphanChildren())
    }

    @Test
    fun `wouldOrphanChildren returns true when selection spans different indent levels with deeper first`() {
        val state = EditorState()
        state.updateFromText("line0\n\tchild\nline2\nline3")
        // Select child + line2 (lines 1-2): first at indent 1, second at indent 0
        state.setSelection(6, 18)
        // First line (indent 1) > shallowest (indent 0) - should warn
        assertTrue(state.wouldOrphanChildren())
    }

    // ==================== moveLinesInternal ====================

    @Test
    fun `moveLinesInternal moves single line up`() {
        val state = EditorState()
        state.updateFromText("line1\nline2\nline3")
        state.focusedLineIndex = 2
        val newRange = state.moveLinesInternal(2..2, 1)
        assertEquals(1..1, newRange)
        assertEquals("line1\nline3\nline2", state.text)
    }

    @Test
    fun `moveLinesInternal moves single line down`() {
        val state = EditorState()
        state.updateFromText("line1\nline2\nline3")
        state.focusedLineIndex = 0
        val newRange = state.moveLinesInternal(0..0, 2)
        assertEquals(1..1, newRange)
        assertEquals("line2\nline1\nline3", state.text)
    }

    @Test
    fun `moveLinesInternal moves block up`() {
        val state = EditorState()
        state.updateFromText("first\nparent\n\tchild\nlast")
        state.focusedLineIndex = 1
        val newRange = state.moveLinesInternal(1..2, 0)
        assertEquals(0..1, newRange)
        assertEquals("parent\n\tchild\nfirst\nlast", state.text)
    }

    @Test
    fun `moveLinesInternal moves block down`() {
        val state = EditorState()
        state.updateFromText("parent\n\tchild\nmiddle\nlast")
        state.focusedLineIndex = 0
        val newRange = state.moveLinesInternal(0..1, 3)
        assertEquals(1..2, newRange)
        assertEquals("middle\nparent\n\tchild\nlast", state.text)
    }

    @Test
    fun `moveLinesInternal returns null for invalid source range`() {
        val state = EditorState()
        state.updateFromText("line1\nline2")
        assertNull(state.moveLinesInternal(-1..0, 1))
        assertNull(state.moveLinesInternal(0..5, 1))
    }

    @Test
    fun `moveLinesInternal returns null for target within source`() {
        val state = EditorState()
        state.updateFromText("line1\nline2\nline3")
        assertNull(state.moveLinesInternal(0..1, 1))
    }

    @Test
    fun `moveLinesInternal adjusts focused line index`() {
        val state = EditorState()
        state.updateFromText("line1\nline2\nline3")
        state.focusedLineIndex = 2
        state.moveLinesInternal(2..2, 0)
        assertEquals(0, state.focusedLineIndex)
    }

    @Test
    fun `moveLinesInternal updates selection when moving selected lines`() {
        val state = EditorState()
        state.updateFromText("line1\nline2\nline3")
        state.setSelection(6, 11) // line2
        state.moveLinesInternal(1..1, 0)
        // Selection should follow the moved line
        assertTrue(state.hasSelection)
    }

    // ==================== Empty line handling ====================

    @Test
    fun `empty lines are treated as indent 0`() {
        val state = EditorState()
        state.updateFromText("\tindented\n\nplain")
        assertEquals(1, state.getIndentLevel(0))
        assertEquals(0, state.getIndentLevel(1))
        assertEquals(0, state.getIndentLevel(2))
    }

    @Test
    fun `move over empty line hops one at a time`() {
        val state = EditorState()
        state.updateFromText("line1\n\nline3")
        state.focusedLineIndex = 2
        val target = state.findMoveUpTarget(2..2)
        assertEquals(1, target) // Empty line is its own group
    }

    // ==================== Hidden indices (showCompleted=false) ====================

    private fun lines(vararg texts: String): List<LineState> = texts.map { LineState(it) }

    @Test
    fun `findMoveUpTarget skips hidden lines`() {
        val l = lines("a", "☑ done", "b")
        val hidden = setOf(1)
        assertEquals(0, MoveTargetFinder.findMoveUpTarget(l, 2..2, hidden))
    }

    @Test
    fun `findMoveUpTarget returns null when target is hidden`() {
        val l = lines("☑ done", "active")
        val hidden = setOf(0)
        assertNull(MoveTargetFinder.findMoveUpTarget(l, 1..1, hidden))
    }

    @Test
    fun `findMoveUpTarget skips hidden block including children`() {
        val l = lines("a", "☑ parent", "\tchild", "b")
        val hidden = setOf(1, 2)
        assertEquals(0, MoveTargetFinder.findMoveUpTarget(l, 3..3, hidden))
    }

    @Test
    fun `findMoveDownTarget skips hidden lines`() {
        val l = lines("a", "☑ done", "b")
        val hidden = setOf(1)
        assertEquals(3, MoveTargetFinder.findMoveDownTarget(l, 0..0, hidden))
    }

    @Test
    fun `findMoveDownTarget returns null when all below are hidden`() {
        val l = lines("active", "☑ done1", "☑ done2")
        val hidden = setOf(1, 2)
        assertNull(MoveTargetFinder.findMoveDownTarget(l, 0..0, hidden))
    }

    @Test
    fun `findMoveDownTarget skips hidden block and finds target`() {
        val l = lines("a", "☑ parent", "\tchild", "b")
        val hidden = setOf(1, 2)
        assertEquals(4, MoveTargetFinder.findMoveDownTarget(l, 0..0, hidden))
    }

    @Test
    fun `findMoveUpTarget skips hidden block above visible target`() {
        // Lines: unchecked, ☑ done1, ☑ done2, empty, unchecked2
        // Moving unchecked2 up should land above the hidden block
        val l = lines("unchecked1", "☑ done1", "☑ done2", "", "unchecked2")
        val hidden = setOf(1, 2)
        assertEquals(1, MoveTargetFinder.findMoveUpTarget(l, 4..4, hidden))
    }

    @Test
    fun `findMoveUpTarget does not skip hidden block at deeper indent`() {
        // Lines: \t☐ parent, \t☑ child (hidden), Root, \t☐ mover
        // Moving mover up should swap with Root, not jump past hidden child
        val l = lines("\t☐ parent", "\t☑ child", "Root", "\t☐ mover")
        val hidden = setOf(1)
        assertEquals(2, MoveTargetFinder.findMoveUpTarget(l, 3..3, hidden))
    }

    @Test
    fun `findMoveUpTarget skips hidden block with children above visible target`() {
        val l = lines("unchecked1", "☑ parent", "\tchild", "", "unchecked2")
        val hidden = setOf(1, 2)
        assertEquals(1, MoveTargetFinder.findMoveUpTarget(l, 4..4, hidden))
    }

    @Test
    fun `findMoveDownTarget stops block expansion at hidden sibling`() {
        // Lines: unchecked1, unchecked2, ☑ done1, ☑ done2, empty
        // Moving unchecked1 down should swap with unchecked2, not jump past hidden lines
        val l = lines("unchecked1", "unchecked2", "☑ done1", "☑ done2", "")
        val hidden = setOf(2, 3)
        assertEquals(2, MoveTargetFinder.findMoveDownTarget(l, 0..0, hidden))
    }

    @Test
    fun `findMoveDownTarget still includes visible children in block expansion`() {
        // Lines: unchecked1, unchecked2, \tchild, ☑ done (hidden), empty
        val l = lines("unchecked1", "unchecked2", "\tchild", "☑ done", "")
        val hidden = setOf(3)
        assertEquals(3, MoveTargetFinder.findMoveDownTarget(l, 0..0, hidden))
    }

    @Test
    fun `getMoveTarget with hiddenIndices move up`() {
        val l = lines("a", "☑ done", "b")
        val hidden = setOf(1)
        assertEquals(0, MoveTargetFinder.getMoveTarget(l, false, EditorState().selection, 2, true, hidden))
    }

    @Test
    fun `getMoveTarget with hiddenIndices move down`() {
        val l = lines("a", "☑ done", "b")
        val hidden = setOf(1)
        assertEquals(3, MoveTargetFinder.getMoveTarget(l, false, EditorState().selection, 0, false, hidden))
    }

    // ==================== Edge cases ====================

    @Test
    fun `single line document disables both move directions`() {
        val state = EditorState()
        state.updateFromText("only line")
        assertNull(state.getMoveTarget(moveUp = true))
        assertNull(state.getMoveTarget(moveUp = false))
    }

    @Test
    fun `move updates cursor to follow moved line`() {
        val state = EditorState()
        state.updateFromText("line1\nline2\nline3")
        state.focusedLineIndex = 0
        state.moveLinesInternal(0..0, 3) // Move to end
        assertEquals(2, state.focusedLineIndex)
    }
}
